src/ratbag_emu/device_handler.py

```diff
@@ -21,6 +21,16 @@ def append_device(device):
         DeviceHandler.devices[DeviceHandler.cur_id] = device
         DeviceHandler.lock.release()
 
+    @staticmethod
+    def wait_for_device_nodes(device_id, timeout=3):
+        start = time()
+        for n, endpoint in DeviceHandler.devices[device_id].endpoints.items():
+            while not endpoint.device_nodes or len(endpoint.device_nodes) == 0:
+                print(f'e.d_n = {endpoint.device_nodes} | len(e.d_n) = {len(endpoint.device_nodes)}')
```

Comment:
doesn't that always print `[] | 0`?

fwiw, in python any 0 length list evaluates as false, so the length check is superfluous. And, for the future, checking a list for False is more efficient than checking the length for zero since in the former case it doesn't have to actually iterate through the elements.

Reply:
This was leftover debug code. I was running to an odd case where this wasn't true. I can't seem to reproduce it right now so I will just remove it. But I remember clearly, only after I added the length check the code started to behave properly.